File: rental-room-management-system-backend/src/main/java/com/rentalroom/management/service/AccountService.java
package com.rentalroom.management.service;

import com.rentalroom.management.common.PageResponse;
import com.rentalroom.management.dto.request.AccountCreateRequest;
import com.rentalroom.management.dto.request.AccountUpdateRequest;
import com.rentalroom.management.dto.response.AccountResponse;
import com.rentalroom.management.entity.Account;
import com.rentalroom.management.enums.Role;
import com.rentalroom.management.exception.BusinessException;
import com.rentalroom.management.repository.AccountRepository;
import jakarta.persistence.criteria.Predicate;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;

@Service
@Transactional
public class AccountService {

    private final AccountRepository accountRepository;
    private final PasswordEncoder passwordEncoder;

    public AccountService(AccountRepository accountRepository, PasswordEncoder passwordEncoder) {
        this.accountRepository = accountRepository;
        this.passwordEncoder = passwordEncoder;
    }

    @Transactional(readOnly = true)
    public PageResponse<AccountResponse> list(Role role, Boolean active, String search, Pageable pageable) {
        Specification<Account> spec = (root, query, cb) -> {
            List<Predicate> predicates = new ArrayList<>();
            if (role != null) {
                predicates.add(cb.equal(root.get("role"), role));
            }
            if (active != null) {
                predicates.add(cb.equal(root.get("active"), active));
            }
            if (search != null && !search.isBlank()) {
                String pattern = "%" + search.toLowerCase() + "%";
                predicates.add(cb.or(
                        cb.like(cb.lower(root.get("fullName")), pattern),
                        cb.like(cb.lower(root.get("username")), pattern)
                ));
            }
            return cb.and(predicates.toArray(new Predicate[0]));
        };
        return PageResponse.from(accountRepository.findAll(spec, pageable).map(AccountResponse::from));
    }

    @Transactional(readOnly = true)
    public AccountResponse get(Long id) {
        return AccountResponse.from(findEntity(id));
    }

    public AccountResponse create(AccountCreateRequest request) {
        if (accountRepository.existsByUsername(request.username())) {
            throw BusinessException.conflict("Tên đăng nhập đã tồn tại");
        }
        Account account = new Account();
        account.setFullName(request.fullName());
        account.setUsername(request.username());
        account.setPasswordHash(passwordEncoder.encode(request.password()));
        account.setRole(request.role());
        account.setActive(true);
        return AccountResponse.from(accountRepository.save(account));
    }

    public AccountResponse update(Long id, AccountUpdateRequest request) {
        Account account = findEntity(id);
        if (!account.getUsername().equals(request.username())
                && accountRepository.existsByUsername(request.username())) {
            throw BusinessException.conflict("Tên đăng nhập đã tồn tại");
        }
        account.setUsername(request.username());
        account.setRole(request.role());
        return AccountResponse.from(accountRepository.save(account));
    }

    /** Accounts not yet linked to any Employee — candidates for the "link existing account" flow on the Employee screen. */
    @Transactional(readOnly = true)
    public List<AccountResponse> listUnassigned() {
        return accountRepository.findUnassigned().stream().map(AccountResponse::from).toList();
    }

    public void changePassword(Long id, String newPassword) {
        Account account = findEntity(id);
        account.setPasswordHash(passwordEncoder.encode(newPassword));
        accountRepository.save(account);
    }

    public AccountResponse setActive(Long id, boolean active) {
        Account account = findEntity(id);
        account.setActive(active);
        return AccountResponse.from(accountRepository.save(account));
    }

    private Account findEntity(Long id) {
        return accountRepository.findById(id)
                .orElseThrow(() -> BusinessException.notFound("Không tìm thấy tài khoản"));
    }
}

File: rental-room-management-system-frontend/src/app/core/services/account.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable, forkJoin, map } from 'rxjs';

import { environment } from '../../../environments/environment';
import { Role } from '../enums/role.enum';
import { ListQuery, PageResponse } from '../models/api-response.model';
import { AccountCreateRequest, AccountResponse, AccountUpdateRequest } from '../models/account.model';
import { buildListParams } from '../utils/http-params.util';
import { ApiService } from './api.service';

export interface AccountListFilter {
  role?: Role | null;
  active?: boolean | null;
  search?: string | null;
}

@Injectable({ providedIn: 'root' })
export class AccountService {
  private readonly api = inject(ApiService);

  list(query: ListQuery, filter: AccountListFilter = {}): Observable<PageResponse<AccountResponse>> {
    const params = buildListParams(query, filter);
    return this.api.get<PageResponse<AccountResponse>>(`${environment.apiUrl}/accounts`, { params });
  }

  get(id: number): Observable<AccountResponse> {
    return this.api.get<AccountResponse>(`${environment.apiUrl}/accounts/${id}`);
  }

  /** Accounts not yet linked to any Employee — for the "link existing account" flow on the Employee screen. */
  listUnassigned(): Observable<AccountResponse[]> {
    return this.api.get<AccountResponse[]>(`${environment.apiUrl}/accounts/unassigned`);
  }

  create(request: AccountCreateRequest): Observable<AccountResponse> {
    return this.api.post<AccountResponse>(`${environment.apiUrl}/accounts`, request);
  }

  update(id: number, request: AccountUpdateRequest): Observable<AccountResponse> {
    return this.api.put<AccountResponse>(`${environment.apiUrl}/accounts/${id}`, request);
  }

  changePassword(id: number, newPassword: string): Observable<void> {
    return this.api.patch<void>(`${environment.apiUrl}/accounts/${id}/password`, { newPassword });
  }

  setActive(id: number, active: boolean): Observable<AccountResponse> {
    return this.api.patch<AccountResponse>(`${environment.apiUrl}/accounts/${id}/active`, { active });
  }

  /** Options list for dropdowns (e.g. the branch-manager picker on the Branch screen). Both admin roles are eligible. */
  listManagerOptions(): Observable<AccountResponse[]> {
    return forkJoin([
      this.list({ page: 0, size: 100 }, { role: Role.ADMIN_TONG }),
      this.list({ page: 0, size: 100 }, { role: Role.ADMIN_CAP_1 }),
    ]).pipe(map(([tong, cap1]) => [...tong.content, ...cap1.content]));
  }
}
